sqlobject/src/main/java/org/jdbi/v3/sqlobject/BatchHandler.java

```diff
@@ -193,20 +193,14 @@ public StatementContext getContext() {
         });
     }
 
-    private Iterator<Object[]> zipArgs(Object[] args) {
+    private Iterator<Object[]> zipArgs(Method method, Object[] args) {
         boolean foundIterator = false;
         List<Iterator<?>> extras = new ArrayList<>();
-        for (final Object arg : args) {
-            if (arg instanceof Iterable) {
-                extras.add(((Iterable<?>) arg).iterator());
-                foundIterator = true;
-            }
-            else if (arg instanceof Iterator) {
-                extras.add((Iterator<?>) arg);
-                foundIterator = true;
-            }
-            else if (arg.getClass().isArray()) {
-                extras.add(Arrays.asList((Object[])arg).iterator());
+        for (int paramIdx = 0; paramIdx < method.getParameterCount(); paramIdx++) {
+            final boolean singleValue = method.getParameters()[paramIdx].isAnnotationPresent(SingleValue.class);
+            final Object arg = args[paramIdx];
+            if (ReflectionArrayIterator.isIterable(arg) && !singleValue) {
```

Comment:
Meh: could swap these two conditions, so that `!singleValue` can short-circuit the `isIterable` call.